alternator/ttl.cc

```diff
@@ -548,13 +550,26 @@ static future<> scan_table_ranges(
             co_return;
         }
         auto units = co_await get_units(page_sem, 1);
-        // We don't to limit page size in number of rows because there is a
-        // builtin limit of the page's size in bytes. Setting this limit to 1
-        // is useful for debugging the paging code with moderate-size data.
+        // We don't need to limit page size in number of rows because there is
+        // a builtin limit of the page's size in bytes. Setting this limit to
+        // 1 is useful for debugging the paging code with moderate-size data.
         uint32_t limit = std::numeric_limits<uint32_t>::max();
-        // FIXME: which timeout?
-        // FIXME: if read times out, need to retry it.
-        std::unique_ptr<cql3::result_set> rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
+        // Read a page, and if that times out, try again after a small sleep.
+        // If we didn't catch the timeout exception, it would cause the scan
+        // be aborted and only be restarted at the next scanning period.
+        std::unique_ptr<cql3::result_set> rs;
+        for (;;) {
+            try {
+                // FIXME: which timeout?
+                rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
```

Comment:
Probably we should have a separate, longer timeout for internal operations. We prefer that the operation be completed slowly than to retry it, and only need the timeout to detect server failures.

In fact infinite timeout should work here, since if we detect a server failure we'll fail any in-progress operations.

Reply:
(not a blocker for this patch)

Reply:
Yes, this is why I had this "FIXME" (the timeout and the comment just moved from pre-existing code), I thought like you but couldn't make up my mind how to define it.

The current executor::default_timeout() is 10 seconds (and can be overridden by the user) is already pretty long. I was surprised that we saw these timeouts in tests - this is a fairly simple query, just one page in an ongoing scan, and I wouldn't have expected it to take 10 seconds and to be honest, I still don't know why they happened.

You're probably right that an "infinite" timeout (or something like it, like 1 hours) should be fine here. But we'll have to verify that it is really fine... For example, imagine that a scan with CL=QUORUM decides to read from replica A (this node) and B, and suddenly B dies. We need to make sure that our code either tries repica C (I think we don't have this feature in scans... just in point queries...), or it immediately returns a timeout. If our code were to wait until the timeout hoping that B comes back to life then we shouldn't use very long timeouts.

Reply:
With lots of tombstones it's easy to timeout, pre e9cbc9ee85c25deb5ba9ee67ffa6e3ca9f904660. It's a critical improvement for anything involving scans. Thanks @denesb !